add instance config to the self-registration page (#102)
neo4j/docs-ops-manager: modules/ROOT/pages/addition/agent-installation/self-registered.adoc

[[register]]
== Register agent

> Agent self-registration omits manually acquiring token information from NOM UI as described in xref:./manual.adoc#register[Manually registering agent].

Reviewer: 👍

Reviewer 2: ^ see above comment

I actually think this is relevant even if in the future we persist more config - the difference between self-registering and manually registered agents is whether the user has to get the token information from the UI or the agent gets it automatically.

[[register]]
== Register agent
Agent self-registration omits manually acquiring token information from NOM UI as described in xref:./manual.adoc#register[Manually registering agent].

> Instead minimal configuration about the NOM server is provided to allow agent to connect and register itself.

Reviewer 3: ```suggestion
Instead, minimal configuration of the NOM server is provided to allow the agent to connect and register itself.
```

> Authorization is granted either through manual approval of the agent in UI or through mutual authentication.

Reviewer 3: ```suggestion
Authorization is granted either through manual approval of the agent in the UI or through mutual authentication.
```

| Port for connecting the agent to the Neo4j log4j appender
| 9500

| `CONFIG_INSTANCE_n_LOG_CONFIG_PATH`
| Path to the instance log4j config. 
If set, appends the appropriate log appender automatically (including the port specified above).
| /var/lib/neo4j/conf/server-logs.xml

| `CONFIG_INSTANCE_n_QUERY_LOG_MIN_DURATION`
| Minimum duration in milliseconds for a query to be logged (optional)
| 100

| `CONFIG_INSTANCE_n_QUERY_LOG_MIN_DURATION_FILTER_ERRORS`
| Enable filter for errors under the minimum duration in milliseconds (optional)
| true

| `CONFIG_INSTANCE_n_QUERY_LOG_DISABLE_OBFUSCATION`
| Disable the string literal obfuscation in queries (optional)
| true
|===

With only the above configuration, agent attempts to connect to NOM server for self-registration which results in agent going into *Standby mode* until authorized.

*Standby mode* is the runtime state of a self registering agent which is not authorized yet to monitor an instance. In this mode agent keeps checking with NOM server for approval status at every preset time interval (__default 30 seconds__).

The agent exits *Standby mode* once it has been manually approved in the UI. 

> === Additional Configuration for mutual authentication

Reviewer: ```suggestion
=== Additional configuration for mutual authentication (optional)
```

It's recommended to set agent name and description if multiple agents are being self-registered on similar hosts as it would lead to confusion with similarly named agents appearing in UI for approval.
====

=== Start configuration

The following environment variables specify start configuration for the agent:
[cols="<,<,<",options="header"]
|===
| Variable
| Description
| Example
| `CONFIG_TLS_TRUSTED_CERTS`
| PEM encoded trusted CA list ()
| `/path/to/a/pem/file`

| `CONFIG_LOG_LEVEL`
| Log level (debug,info,warn,error)
| info

| `CONFIG_LOG_FILE`
| Path to the log file
| `/var/log/nom-agent/log.txt`
|===

[NOTE]
====
Since agent-server communication needs to be encrypted, you need to configure the agent so that it trusts the server's certificates.
The file that contains the trusted certificate list (PEM encoded) can be specified through the `CONFIG_TLS_TRUSTED_CERTS` environment variable.

> While most operating systems default to the system-wide trusted certificates, that's not the case on Windows.

Reviewer 3: ```suggestion
While most operating systems default to the system-wide trusted certificates, it is not the case on Windows.
```

> For that reason, you **must** set this environment variable on Windows.

Reviewer 3: ```suggestion
For this reason, you **must** set this environment variable on Windows.
```

====

=== Instance configuration
For each managed DBMS instance on the host, the following environment variables need to be set to allow the agent to access the instance:

[cols="<,<,<",options="header"]
|===
| Variable
| Description
| Example

| `CONFIG_INSTANCE_n_NAME`
| Name of nth instance
| my-instance-n

| `CONFIG_INSTANCE_n_BOLT_URI`
| Bolt URI for nth instance with bolt or bolt+s protocol
| bolt://localhost:7687 or bolt+s://localhost:7687 or bolt+ssc://localhost:7687, depending on the local database setup

| `CONFIG_INSTANCE_n_BOLT_USERNAME`
| Bolt user name for nth instance
| neo4j

| `CONFIG_INSTANCE_n_BOLT_PASSWORD`
| Bolt password for nth instance
| password
|===

[[querylog]]

> To enable the query log collection functionality, there are also a few additional environment variables.

Reviewer 3: ```suggestion
To enable the query log collection functionality, there are a few additional environment variables.
```

| 9500

| `CONFIG_INSTANCE_n_LOG_CONFIG_PATH`
| Path to the instance log4j config. 
If set, appends the appropriate log appender automatically (including the port specified above).
| /var/lib/neo4j/conf/server-logs.xml

| `CONFIG_INSTANCE_n_QUERY_LOG_MIN_DURATION`
| Minimum duration in milliseconds for a query to be logged (optional)
| 100

| `CONFIG_INSTANCE_n_QUERY_LOG_MIN_DURATION_FILTER_ERRORS`
| Enable filter for errors under the minimum duration in milliseconds (optional)
| true

| `CONFIG_INSTANCE_n_QUERY_LOG_DISABLE_OBFUSCATION`
| Disable the string literal obfuscation in queries (optional)
| true
|===

With only the above configuration, the agent will connect to NOM server, register itself and then enter *Standby mode*.

*Standby mode* is the runtime state of a self-registering agent which is not authorized yet to monitor instances. 
In this mode agent keeps checking with NOM server for approval status at every preset time interval (__default 30 seconds__).

The agent exits *Standby mode* once it has been manually approved in the UI. 

=== Additional configuration for mutual authentication (optional)

Mutual authentication eliminates the manual approval step. 

> NOM server will immediately authorize an agent that registers using a trusted certficate.

Reviewer 3: ```suggestion
The NOM server immediately authorizes an agent that registers using a trusted certificate.
```

The following configuration is required to enable mutual authentication:

[cols="<,<,<",options="header"]
|===
| Variable
| Description
| Example

| `CONFIG_CLIENT_CERT`
| PEM encoded Agent certificate for mutual TLS
| `/path/to/a/pem/file`

| `CONFIG_CLIENT_KEY`
| PEM encoded Agent key for mutual TLS
| `/path/to/a/pem/file`
|===

> IMPORTANT: *In addition to the above configuration, NOM server also needs to be configured to trust the agent certificates as described xref:/installation/server.adoc#config_ref[here]*

Reviewer 3: ```suggestion
[IMPORTANT]
====
In addition to the above configuration, the NOM server also needs to be configured to trust the agent certificates as described xref:/installation/server.adoc#config_ref[here].
====
```

> NOTE: Self-signed certificates for agents in test and demo envrionments can be generated as documented xref:./agent-self-signed-certificates.adoc[here]

Reviewer 3: ```suggestion
[NOTE]
====
Self-signed certificates for agents in test and demo environments can be generated as documented xref:./agent-self-signed-certificates.adoc[here].
====
```

[[running-agent]]
== Run agent

> To run a self-registering agent an additional command line option is provided as `-s` for short and `--self-register`.

Reviewer 3: ```suggestion
To run a self-registering agent, an additional command line option is provided as `-s` for short and `--self-register`.
```

== Verify agent setup
Ensure agent has contacted NOM server, is online and is reporting DBMS(s) correctly.

. Return to Agents listing in global settings.
+
image::agents.png[width=800]
. Find self-registered agent in list.
 ** If the agent is not in the list then go back to where the agent is running and check the logs.
    It may be that the server address is configured incorrectly or the TLS certificates are not correctly specified. 

. Approve agent if required.
** An agent in standby mode shows up in the list of agents in NOM UI with `Unauthorized` status. 
+
image::agent-unauthorized.png[width=800]

> To enable the agent to continue its normal execution, agent needs to be approved from NOM UI as shown below:

Reviewer 3: ```suggestion
To enable the agent to continue its normal execution, the agent needs to be approved from the NOM UI as shown below:
```

. Return to Agents listing in global settings
+
image::agents.png[width=800]
. An agent in standby mode shows up in the list of agents in NOM UI with `Unauthorized` status. 

> To enable the agent to continue it's normal execution, agent needs to be approved from NOM UI as shown below:

Reviewer 2: ```suggestion
To enable the agent to continue its normal execution, agent needs to be approved from NOM UI as shown below:
```

Reviewer 2: I misspelled it in many places I guess.

+
NOTE: An unauthorized self-registering agent will have a default name derived from its host information. It's recommended to update this name or use the optional configuration described <<agent-meta-data,here>>.
+
image::agent-approve.png[width=800]
. Once the agent is approved, the agent will receive initial configuration from NOM server which it persists on the instance host in a file named `nom-agent-config.yaml`.
. Upon approval, the agent status changes to `Offline` until the agent connects to the server with the initial configuration.
. Once the agent is approved (through UI or through mutual authentication), the agent will receive token information from NOM server which it persists on the instance host in a file named `nom-agent-config.yaml`.
. Upon approval, the agent status changes to `Offline` until the agent connects to the server with the token information.
+
image::agent-approved-offline.png[width=800]
. Once the agent is successfully connected to Server and reports the instances it needs to monitor, it's status changes to `Online`.
+
image::agent-approved-online.png[width=800]
. Check that there is a value for _Last contact time_.

>     ** If the agent has never contacted NOM server then go back to where the agent is running and check the logs.

Reviewer 2: I know that these are the same steps for manual installation, but they still apply to self-registration so kept them as they were. This is the duplication of docs I was referring to.

Author: I actually removed them because they made no sense if the user has just done the manual approval steps. I couldn't quite work out how best to handle the mutual authentication case elegantly...

Author: Though on re-reading 
```
 Verify that the agent has a current status of `Online`. 
    ** If the agent is not currently online then go to where it is running and check the logs.
```

is probably still relevant. 

and maybe the part about 'last contact time' just needs to be moved up and changed to 'if you can't see the agent'.

Author: I'll try again 😄
